wallet/src/key_chain/account_key_chain.rs

```diff
@@ -0,0 +1,262 @@
+        purpose: KeyPurpose,
+    ) -> KeyChainResult<ExtendedPublicKey> {
+        self.get_leaf_key_chain_mut(purpose).issue_key(db_tx)
+    }
+
+    /// Get the private key that corresponds to the provided public key
+    #[allow(dead_code)] // TODO remove
+    pub fn get_private_key(
+        &self,
+        parent_key: &ExtendedPrivateKey,
+        requested_key: &ExtendedPublicKey,
+    ) -> KeyChainResult<ExtendedPrivateKey> {
+        let derived_key =
+            parent_key.clone().derive_absolute_path(requested_key.get_derivation_path())?;
+        if &derived_key.to_public_key() == requested_key {
+            Ok(derived_key)
+        } else {
+            Err(KeyChainError::KeysNotInSameHierarchy)
+        }
+    }
+
+    /// Get the leaf key chain for a particular key purpose
+    pub fn get_leaf_key_chain(&self, purpose: KeyPurpose) -> &LeafKeyChain {
+        match purpose {
+            KeyPurpose::ReceiveFunds => &self.receiving_key_chain,
+            KeyPurpose::Change => &self.change_key_chain,
+        }
+    }
+
+    /// Get the mutable leaf key chain for a particular key purpose. This is used internally with
+    /// database persistence done externally.
+    fn get_leaf_key_chain_mut(&mut self, purpose: KeyPurpose) -> &mut LeafKeyChain {
+        match purpose {
+            KeyPurpose::ReceiveFunds => &mut self.receiving_key_chain,
+            KeyPurpose::Change => &mut self.change_key_chain,
+        }
+    }
+
+    // Return true if the provided public key belongs to this key chain
+    pub fn is_public_key_mine(&self, public_key: &PublicKey) -> bool {
+        for purpose in KeyPurpose::ALL {
+            if self.get_leaf_key_chain(purpose).is_public_key_mine(public_key) {
+                return true; // Return early to avoid checking all leaf key chains
+            }
+        }
+        false
+    }
+
+    // Return true if the provided public key hash belongs to this key chain
+    pub fn is_public_key_hash_mine(&self, pubkey_hash: &PublicKeyHash) -> bool {
+        for purpose in KeyPurpose::ALL {
+            if self.get_leaf_key_chain(purpose).is_public_key_hash_mine(pubkey_hash) {
+                return true; // Return early to avoid checking all leaf key chains
+            }
+        }
+        false
+    }
+
+    /// Derive addresses until there are lookahead unused ones
+    pub fn top_up_all<B: Backend>(&mut self, db_tx: &mut StoreTxRw<B>) -> KeyChainResult<()> {
+        for purpose in KeyPurpose::ALL {
+            self.get_leaf_key_chain_mut(purpose).top_up(db_tx)?;
+        }
+        Ok(())
+    }
+
+    pub fn get_account_info(&self) -> AccountInfo {
+        AccountInfo::Deterministic(DeterministicAccountInfo::new(
+            self.root_hierarchy_key.clone(),
+            self.account_pubkey.clone(),
+            self.lookahead_size,
+        ))
+    }
+
+    pub fn get_lookahead_size(&self) -> u32 {
+        self.lookahead_size
+    }
+
+    pub fn set_lookahead_size<B: Backend>(
```

Comment:
Shouldn't this now be constant?

My understanding is that bitcoin core wouldn't allow this because you can't guarantee that you haven't used pre-generated addresses when you increase the gap.

Reply:
But we don't handle the case of reducing the size properly, right? Why not modify that test to make the assumption that this doesn't change?

If the value won't change, then there's no need to test that it'll ever change.